pkg/common/functions.go

```diff
@@ -0,0 +1,62 @@
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+)
+
+// CheckContext checks if a go context is still active or done
+// When it's done, returns a generic error
+//
+// func myFunc(ctx context.Context) error {
+//   if err := common.CheckContext(ctx); err != nil {
+//   	 return err
+//   } else {
+//     doSomething()
+//   }
+// }
+func CheckContext(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("Context aborted")
```

Comment:
Probably we will want an scenario where the context was cancelled and it didn't return an error, the cancellation was intended, part of the execution.

Reply:
It wouldn't make any sense to use this helper function in such case, I think. However, you can still just ignore the response, no?

```go
  if err := common.CheckContext(ctx); err != nil {
    // Yeah, it looks like context is over
    // Do something cool then
  }
```

Reply:
I think it's not worth the abstraction.. specially when we most probably we'll take different actions when it's cancelled. Maybe @eguzki can enlighten us ?

Reply:
If there will be different actions when it is cancelled,  I would do something like this

```
func IsContextDone(ctx context.Context) bool {
	select {
	case <-ctx.Done():
		return true
	default:
		return false
	}
}
```
But still, this approach might not be valid for all scenarios. In fact, the most common scenario would be the one documented in their [source code](https://golang.org/src/context/context.go), wait  for many things to happen and if the context is cancelled first, then react:
```
func Stream(ctx context.Context, out chan<- Value) error {
	for {
		v, err := DoSomething(ctx)
		if err != nil {
			return err
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case out <- v:
		}
	}
}
```

Reply:
Guys, I agree replacing

```ruby
		return fmt.Errorf("Context aborted")
```

with
```ruby
		return ctx.Err()
```

But I think having the helper method still makes sense. A few things to have in mind:
1. I wouldn't want to have to write a `select..case ` statement every time for something that a simple `if` would do.
2. Most of the places where it's used (if not all of them!) returns an error to the caller. Making this check to also return an error is an easy way to avoid dozens of different error messages, spread all over the code, for something that is essentially the same. So, I wouldn't make it return a `bool`.
3. The example of usage [pointed out](https://github.com/3scale-labs/authorino/pull/49#discussion_r584582801) by @eguzki is one where, _first_, the code does something (potentially something heavy), then if the context is not done yet, uses the result of that something; otherwise, returns the context error. Maybe that's because it is inside of a `for` loop and trying to depict that the task would usually be performed at least once before the context could be cancelled. Here, however, I mostly implemented the check _before_ running the (potentially heavy) task, not _after_ it. I guess there's no ideal solution for this type of problem actually. If it's after the task, it means all concurrent jobs will do something and some of them later realize it wasn't needed to spend time on it after all; if it's before, then there is the risk of, due to concurrency, all jobs to misjudge that performing the task is needed, when actually it isn't or it won't be soon but you cannot tell now. Hopefully, this difference won't matter much because we are adding here not one single check per evaluator, per possibly several of them, interspersed in middle of multiple steps performed by each evaluator.

Reply:
I'm not totally convinced but I'll keep an open mind and see the future usage of the helper. Let's roll with it

Reply:
> Here, however, I mostly implemented the check before running the (potentially heavy) task, not after it.

That depends on the "cancelling" approach you want. Let's say you want to do something that takes seconds or minutes. In heavy load, do you want to have the long running task finished before returning back to handle the next job? That is one way. Another way is to cancel the long running job as soon as the cancel signal arrives. For that approach, you need the example of usage pointed out. 

Anyway, as said earlier, I think it is to soon to have this kind of concerns. They can be tackled later on.